### PR TITLE
Reduce conditional_latest_ready_time pessimism

### PR DESCRIPTION
**Pull request by knokko**
Before this commit, the `conditional_latest_ready_time` function was needlessly pessimistic because predecessor jobs could postpone the latest ready time of `j_high`, even if they are guaranteed to be finished before `j_low` is ready.

We already had a similar optimization for single-core analysis, which is now generalized to multiple cores.

### CODE REVIEW

File: include/global/state_space_data.hpp [hunk 1]
```diff
@@ -198,6 +201,14 @@ namespace NP {
 						// since we assumed that all predecessors of `j_high` were already dispatched.
 						if (num_cpus == 1) continue;
 
+						// The optimization above can be generalized to multiple cores: when
+						// - `j_pred` is occupying the next core that will become available, and
+						// - the second core cannot become available until `j_pred` has finished
+						// we know that `j_pred` is certainly finished when `j_low` becomes ready.
+						if (ncores == 1 && ft.min() == s.core_availability(1).min() &&
```
Comment (gnelissen) [inline include/global/state_space_data.hpp:208]: This tests is not well supported. I would need a formal proof to accept it. I also feel it is overly complicated. Why the following test is not enough? 
`if( ncores==1 and ft.max() < s.core_availability(2).min())`
Note that I used a strict inequality as I am unsure of the corner case.
I would expect that, if a predecessor is certainly finished before two cores may become available, then it either completed before the first core becomes available, or it executes on the first core that becomes available. In either case, it must be finished when the first core becomes available.

**Review by gnelissen — COMMENTED** [review]
A formal proof that it is safe is requried.

File: include/global/state_space_data.hpp (hunk 1 above)
Comment (knokko) [reply include/global/state_space_data.hpp:208]: Ok, let's address these points one by one:

_Why the following test is not enough?_
`if( ncores==1 and ft.max() < s.core_availability(2).min())`

This would indeed be enough. I think the condition
`if( ncores==1 and ft.max() <= s.core_availability(2).min())` is equivalent to
`if (ncores == 1 && ft.min() == s.core_availability(1).min() && ft.max() == s.core_availability(1).max() && ft.max() <= s.core_availability(2).min()`
I simply find it easier to reason about my longer condition.

_then it either completed before the first core becomes available_
I think it is impossible that `j_pred` is completed before the first core becomes available, since a core will become available when `j_pred` completes. Unless of course a different job already started running on its core... I'm no longer so sure about this... perhaps I am indeed being needlessly pessimistic...

_Note that I used a strict inequality as I am unsure of the corner case._
Ok, let's consider this corner case. Let `ca(n)` denote `s.core_availability(n)`. By definition of `ca`, we know that `ca(1).min() <= ca(2).min() <= ca(2).max()` and `ca(1).max() <= ca(2).max()`. When `ft.max() == ca(2).min()`, there exists some other job `j'` that may finish at the same time as `j_pred`, or later. When `j'` finishes later, it is not important.

When `j'` finishes at the same time as `j_pred`, 2 cores become available at the same time when `j'` finishes. In this situation, it is possible that `j_low` starts at the same time as `j_high`, and it is possible that `j_high` starts at the same time as some other unrelated job. The latter case is not important since `j_low` is not dispatched next. In the former case, the SAG can either model the dispatch or `j_low` first, or model the dispatch of `j_high` first. Since they are dispatched at the same time, the resulting state is the same. But, modeling the dispatch of `j_high` first is most convenient for us, so let's disregard `j_low`.

_This tests is not well supported. I would need a formal proof to accept it._
Just like the other disregard/`continue` conditions, we disregard `j_pred` when it is impossible that `j_pred` postpones the ready time of `j_high` to a time instant later than the time at which `j_low` can start. Proving this will be complicated, but I think we already agree about this. So I will prove only the following statement:
**When j_pred.finish_time.max() <= s.core_availability(2).min(), it is impossible that j_pred postpones the ready time of j_high to a time instant later than the time at which j_low can start**

If `j_pred` is already finished, it clearly can't postpone the ready time of `j_high`. If `j_pred` is still running, we know that a core will become available when `j_pred` finishes, thus `s.core_availability(1).max() <= j_pred.finish_time.max()`. If also `j_pred.finish_time.max() <= s.core_availability(2).min()`, it follows from transitivity that `s.core_availability(1).max() <= s.core_availability(2).min()` hence `j_pred` is certainly the next job that will finish (although another job might finish at exactly the same time), and all cores will be occupied until `j_pred` is finished.

Since all cores are occupied until `j_pred` is finished, there is no way that `j_low` can start before `j_pred` is finished. And when `j_pred` is finished, it can no longer postpone the ready time of `j_high` since there is no suspension between `j_pred` and `j_high`.

Comment (knokko) [reply include/global/state_space_data.hpp:208]: In hindsight, you were right about the strict inequality
*Note that I used a strict inequality as I am unsure of the corner case.*

I constructed a counterexample where the inequality really matters, and will push it soon.

Comment (knokko) [reply include/global/state_space_data.hpp]: The invalid reasoning in my broken proof was *, and all cores will be occupied until j_pred is finished.*

My initial version disregarded the possibility that `j_pred` was running, but another core was free. So the case was this:
- core_availability(1) = [0, 0]
- core_availability(2) = [ft.min(), ft.max()]

When `ft.min() == ft.max()`, this would indeed be unsound, but it would still work if `ft.min() < ft.max()` since then `ft.max() > core_availability(2).min() == ft.min()`. In my updated version, this is prevented by the additional `&& ft.min() < s.core_availability(2).min()` check.

Alternatively, I could use my old `ft.min() == core_availability(1).min() && ft.max() == core_availability(1).max()` checks, but I'm not sure which one is less pessimistic.

**Comment by knokko** [comment]
@gnelissen I found a much better way to prove this, and made the check tighter. I propose the following improvement:
`if (ncores == 1 && ft.min() <= s.core_availability(1).min() && ft.max() <= s.core_availability(2).min()) continue;`

Furthermore, I propose to rename `ncores` to `required_cores` since I find `ncores` very easy to confuse with `num_cpus`.

The soundness proof of my improvement uses 4 claims:

(1) When j_pred cannot postpone the ready time of j_high to a time instant *later than* the moment j_low can start, we can safely disregard j_pred.
(2) j_low cannot start until at least 1 core is available.
(3) So if all cores are certainly occupied until j_pred is finished, we can disregard j_pred.

I think the first 3 claims don't need a proof, so I only prove the fourth claim.

(4) If ft(j_pred).min() <= ca(1).min && ft(j_pred).max() <= ca(2).min() then no core can be available before j_pred is finished.

ft(j) denotes finish time of j and ca(n) denotes core_availability(n).

Proof:
Assume for a contradiction that a core becomes available at time T before j_pred is finished. Then, since j_pred is not finished yet, another core could become available at ft(j_pred).min() thus ca(2).min() <= ft(j_pred).min() hence ca(2).min() <= ft(j_pred).max().

From the condition ft(j_pred).max() <= ca(2).min(), it follows that ft(j_pred).max() == ca(2).min() == ft(j_pred).min(), so j_pred finishes exactly at ca(2).min().

Since another core became available *before* j_pred was finished, it follows that T < ca(2).min() thus T == ca(1).min() < ca(2).min() == ft(j_pred).min(), which contradicts the condition that ft(j_pred).min() <= ca(1).min().

**Comment by gnelissen** [comment]
@knokko it is much more convincing. Some comments:

> I propose to rename ncores to required_cores

OK. Though, if we change the name, I would be more specific yet and specify in the name that it is the cores required by j_low (`j_low_required_cores`)

The proof contains a few approximations, but seems to go in the correct direction.
> since j_pred is not finished yet, another core could become available at ft(j_pred).min() thus ca(2).min() <= ft(j_pred).min() hence ca(2).min() <= ft(j_pred).max().

I do not think that it is true. However, you could rewrite that part of the proof as follows:
Define F as being the finishing time of j_pred. By your hypothesis, T<F. By definition of ca(1).min() and ca(1)max(), ca(1).min()<= T <=ca(1).max(), and by definition of ft.min and ft.max, you have  ft(j_pred).min() <= F <= ft(j_pred).max().
Since j_pred is not finished yet when the first core become available at time T, a second core must become available at or before F. Since ca(2).min() is the earliest time a second core becomes available in any execution scenario, we have ca(2).min() <= F <= ft(j_pred).max().

> so j_pred finishes exactly at ca(2).min().

I would replace it by 
ft(j_pred).min() = ft(j_pred).max() = F.

> Since another core became available before j_pred was finished, it follows that T < ca(2).min() thus T == ca(1).min() < ca(2).min() == ft(j_pred).min()

It is simpler to remind that ca(1).min()<= T and that by the initial hypothesis, T<F <=> ca(1).min()<F=ft(j_pred).min(), which contradict that ft(j_pred).min() <= ca(1).min().


Now, according to the proof, it seems to me that the condition `ft(j_pred).min() <= ca(1).min()` is only required to handle the case were `ft.max() = s.core_availability(2).min()`. If `ft.max() < s.core_availability(2).min()`, then we do not need to enforce `ft(j_pred).min() <= ca(1).min()` because we already reached a contradiction as soon as we proved that the hypothesis required that ca(2).min() <= F <= ft(j_pred).max(). Based on this, I would propose the following, improved test:

```
if (ncores == 1 && (ft.max() < s.core_availability(2).min() || (ft.max() == s.core_availability(2).min() && ft.min() <= s.core_availability(1).min()))
    continue;
```

**Comment by gnelissen** [comment]
another thing. Why do we need `ncores == 1`? The proof does not use it.

**Comment by knokko** [comment]
As you wish
